### ISSUE TITLE
Multiple copies of the same model don't work

### ISSUE THREAD

**Issue by nm156** [Pugnator/openvsm#11]
See branch patch-1 testcases in my fork.  It contains a V7 schematic and a simple multi-part device (7408) script.

Only 1 part of a multi-part device simulates at a time even across devices
Not sure if I did something wrong with the model script or there is an issue with openVSM.

Did not want to pull this if it is my fault.

**Comment by Pugnator** [comment]
First I want to thank you so much for your contribution!
I will check it out at home today

**Comment by Pugnator** [comment]
I didn't test it with multipart, so it is nice occasion to test with

**Comment by nm156** [comment]
I think there is a deeper rooted issue with openvsm.   Take your 'sample' design and 'block copy' and 'block paste'  the entire schematic. Then run it.  The first copy will cease to function, and the second one will simulate.  This is the same effect I am seeing with multi-part.  It appears that only one appearance of any openvsm device may be simulated on a design at a given time.

I'll help out as much as I can. I'm new to Lua and not a great programmer.

I plan to design a simple 24 hour clock with by re-creating some TTL devices (7490, 7447 etc)  as an excercise learning openvsm and Lua.  The most simple was the AND gate so I started there.

**Comment by nm156** [comment]
I'm taking a wild guess at this, but it seems to me that there is only one global "lua_State* luactx in vsm_api.c that is used by any instance of the DLL with the same name. (So each call clobbers the previous until last caller wins)
This is probably a side effect of the C++ to C conversion of the API.  (C++ would have a *new* *class* with an isolated state and other 'global' variables each instantiation) Somehow a new state needs set up in DllMain that keeps the Lua instance and any global variables separate per use.  From what I read onlne, the lua_State itself is self contained. [Stack overflow](/http://stackoverflow.com/questions/3018434/embedding-multiple-lua-instances-in-a-multiple-threaded-program)

I could be totally wrong on it though...

**Comment by Pugnator** [comment]
As it all the same DLL, the same instance of Lua virtual machine is used

**Comment by nm156** [comment]
I hacked together a test case for multiple Lua instances summarized as below

    struct lua_instance
    {
    	lua_State* luactx;
    	VSM_PIN device_pins[32];
    };
    
    unsigned int dll_instance = -1;
    struct lua_instance lua_instances[32];
    
    /* Init Lua */
    lua_instances[dll_instance].luactx = luaL_newstate();
    
    bool APIENTRY
    DllMain ( HINSTANCE hInstDLL, uint32_t fdwReason, LPVOID lpvReserved )
    {
    	( void ) hInstDLL;
    	( void ) fdwReason;
    	( void ) lpvReserved;
    
    	dll_instance++;
    	
    	return true;
    }

it compiles and works with one device, but the same as before with two.  I am beginning to suspect this problem is rooted with the Proteus interface as opposed to the Lua instance issue.

**Comment by Pugnator** [comment]
There are also several global instances expept lua state and pin holder
IINSTANCE* model_instance; ///< Global model instance object
IDSIMCKT* model_dsim; ///< Global DSIM object

I think exactly like you. Somehow identify current DLL instance and use a struct array for handles

**Comment by Pugnator** [comment]
As I found out in C you can't do nothing with except to rename DLL for each device instance in background, but it is impossible via VSM API. But seems I found much better solution inside VSM API.
Working on it

**Comment by nm156** [comment]
Found an interesting VSM interface at
http://purebasic.mybb.ru/viewtopic.php?id=496

Looks similiar to the interface type you are using with virtual tables. Of particular interest is how new instances are created and changed in the virtual table.

    Procedure ModelNEW()
      Protected *this.ClassParam=0
  
      *this = AllocateMemory(SizeOf(ClassParam)) 
      If *this
        ;InitializeStructure(*this, ClassParam)
        *this\vt = ?ClassFunct
      EndIf   
  
      ProcedureReturn *this 
    EndProcedure

I downloaded this and tried the example. I correctly simulates multiple devices on the design.

**Comment by Pugnator** [comment]
Yes, it is the main idea - separate memory for each device instance. At the moment I extend Proteus objects to include openvsm data and hold it itself. It will fix all the problems with multiple instances.
And furthermore - it is much better engine design at all

**Comment by Pugnator** [comment]
OK, here is the result of investigation. 
The main problem was the fact of each engine instance uses static and global variables.
When Proteus issued new device creation/initialization a static object was returned. But it is wrong design at all, but I forgot about it as initially I was planning to use it for myself as CPU model builder and therefore not planned to use it with multiple devices.
Now I repaired this disease.
I had to rewrite or edit the whole code and now it is much more correct Linux to MS VC++ ABI translator than it was before.

It is fixed by extending model object to include all data we need, and allocate it each time model is created and save all as pointers there. Proteus know nothing about it and it doesn't matter.

Also I fixed several very bad bugs (ans possibly create several new).

It is partly works now and I need some more time to fix issues. But seems multiple instances works fine

**Comment by nm156** [comment]
Excellent news and great work!  One side effect of the rewrite might be that now a single DLL named openvsm.dll could be used for any device being modeled.  There might be a field somewhere in Proteus (or a custom one called SCRIPT) we could use to assign the Lua script to the model.

**Comment by Pugnator** [comment]
Yes, may be script should be named after the device name? But as a name may contain special symbols or spaces (which potentially can be treated wrong) better create separate string parameter

**Comment by nm156** [comment]
I tested this:

    {*DEVICE}
    NAME=TEST_PART
    {PREFIX=U}
    {*PROPDEFS}
    {PRIMITIVE="Primitive Type",HIDDEN STRING}
    {MODDLL="VSM Model DLL",READONLY STRING}
    {VSM_SCRIPT=Program File,FILENAME,FALSE,,Lua VSM Script Files/*.lua  <<---- file chooser
    {PACKAGE="PCB Package",PACKAGE,2,DIL14,SO14}
    {*INDEX}
    {DESC=SAMPLE DLL}
    {CAT=TTL 74 series}
    {SUBCAT=Gates & Inverters}
    {ITFMOD=}
    {MFR=}
    {*COMPONENT}
    {PRIMITIVE=DIGITAL}
    {MODDLL=sample.dll}  <<---- openvsm.dll (always?)
    {VSM_SCRIPT=sample.lua}  <<---- default script name (same as device name?)
    {PACKAGE=DIL14}
    *PINOUT DIL14
    {ELEMENTS=1}
    {PIN "Pin1" = 1}
    {PIN "Pin2" = 2}
    {PIN "Pin3" = 3}
    {PIN "Pin4" = 4}
    
    *PINOUT SO14
    {ELEMENTS=1}
    {PIN "Pin1" = 1}
    {PIN "Pin2" = 2}
    {PIN "Pin3" = 3}
    {PIN "Pin4" = 4}

and it looks like it will work. "SCRIPT" is already reserved in Proteus so I used "VSM_SCRIPT".  I put the DLL in the MODELS directory and the script with the project design.  I'm not sure if the "LUAVSM" variable is even needed anymore  if the code is modified. 

 It also appears that the current code is statically linked to the Lua library, so the lua53.dll is not needed.

**Comment by Pugnator** [comment]
If we add filechooser, environment variable can be removed, but is it really handy to choose script for every device? The problem is we can't say for sure where exactly DLL is located as only main process full path is always available from inside the program.
Always put it alongside with the project? And what if you create new project - duplicate the script?
About linking - yes, it is the rudiment of early design.. will remove this dependency from README.
I will push test version today. I'm still testing/fixing new code.

**Comment by nm156** [comment]
Agree. I was thinking more for development than production.

On an unrelated note, I noticed this in the current code:

    lua_Number argnum = lua_gettop ( L );

which always returns '0'. So

    if ( 1 > argnum )
    {
        out_error ( model, "Function %s expects 1 arguments got %d\n", __FUNCTION__ , argnum );
        return 0;
    }

never executes to catch parameter errors.

Shouldn't the code be:

    int argnum = lua_gettop ( L );

That correctly shows the args when I tested.

**Comment by Pugnator** [comment]
Yes, arg check is already fixed, thank for noticing about it!
set/set pin_bool is fixed too

**Comment by Pugnator** [comment]
Please check
https://github.com/Pugnator/openvsm/tree/bug%2311
It works but still in testing

**Comment by Pugnator** [comment]
Script should be named after the LUA string parameter

**Comment by Pugnator** [comment]
Please note that Lua syntax is changed, now you have to pass the model object in most cases.
I will provide a new Sample design soon

**Comment by nm156** [comment]
Yep. Caught that when my scripts were blowing up :)

Wouldn't it be better to keep the model pointer out of the script? (and invisible)

Do something like this once:

    // register literal __this to model pointer
    lua_pushliteral(luactx, "__this");
    lua_pushlightuserdata(luactx, this);
    lua_settable(luactx, LUA_REGISTRYINDEX);


Then access the model from Lua by:

    void* lua_get_model( lua_State* L )
    {
       void* model;

       lua_pushliteral(L, "__this");
       lua_gettable(L, LUA_REGISTRYINDEX);
       model = lua_touserdata(L, -1);
       return model;
    }

In lua_xxx functions get the model from the registry by

    void* model = lua_get_model(L);

and cast model as necessary to c function

**Comment by Pugnator** [comment]
Wow, nice trick ;)

**Comment by nm156** [comment]
I tested the registry model on one function, and the method works.

    int lua_print_warning ( lua_State* L )
    {
    	int argnum = lua_gettop ( L );  // using int instead of lua_number
    	const char* text = lua_tostring ( L, -1 );
    	void* model = lua_get_model(L);
    	
       if ( 1 < argnum )   // <---- NOTE in the master branch a lot of comparisons incorrectly used ">"
    	{
    		out_error ( model, "Function %s expects 1 arguments got %d\n", __FUNCTION__ , argnum );
    		return 0;
    	}
    	out_warning ( model, text );
    	return 0;
    }

The lua script 

    function device_init()  
    	warning("Lua script called warning")
    	warning("Lua script called warning", "2nd ARG")
    	warning("Lua script called warning", "2nd ARG", "3rd ARG")
    end

Results

![log](https://cloud.githubusercontent.com/assets/1387648/6239656/da98fb02-b6d3-11e4-86f4-556d072cf377.png)

Type checking is another matter to resolve.

The only quirk is that all of the expected args need to be pulled first before calling lua_get_model().
There might be a fix for that - I know just enough now to be dangerous.

.

**Comment by Pugnator** [comment]
I think it is not a Lua way to fail on passing more argument than needed. They should be ignored. But when there are too few arguments - it is a really bad situation that probably can lead to segfault.

**Comment by nm156** [comment]
On another note, I decided to try LuaJIT and see if it would work.  One minor function to add and it worked "out of the box" at least on simple test cases.

    // --------------------------------------------------------
    // Lua helper functions
    // Use with Lua 2.0.3 JIT
    // https://github.com/keplerproject/lua-compat-5.2
    // compatibilty function
    // --------------------------------------------------------
    #if (LUA_VERSION_NUM == 501)
    
    extern "C" void lua_len (lua_State *L, int i) {
      switch (lua_type(L, i)) {
        case LUA_TSTRING: // fall through 
        case LUA_TTABLE:
          if (!luaL_callmeta(L, i, "__len"))
            lua_pushnumber(L, (int)lua_objlen(L, i));
          break;
        case LUA_TUSERDATA:
          if (luaL_callmeta(L, i, "__len"))
            break;
          // maybe fall through 
        default:
          luaL_error(L, "attempt to get length of a %s value",
                     lua_typename(L, lua_type(L, i)));
      }
    }
    #endif

I'm all over the board with ideas right now, treating the code base as a playground rather than something I can merge back.  I think once a stable foundation is in place I can start pushing changes.

**Comment by Pugnator** [comment]
LuaJIT is fine thing, I thought about it but didn't test and not sure if there will be any performance boost in this case. Need benchmark device =)

**Comment by Pugnator** [comment]
Added stack check as model window. As I'm not sure if wrong parameter popped from the stack will not lead to segfault.

**Comment by nm156** [comment]
Nice job on the stack check!  It should probably have type checking too to make it bulletproof, which leads into my question.

One thing that really bothers me about Lua is the dynamically typed aspect.

I've spent a lot of time chasing down mispellings of functions and an
extensive google search hasn't brought me any closer to a final solution.
 
For example (somewhat exaggerated):
      
    set_pin_state(Pin3, SHI)
    warning("Lua script called warning")
    set_pins_state(Pin3, SHI)	
    blah("TEST", 1)
    foo(bar)

All execute, the last 3 without any indication that nothing is happening.

I found this piece of code in Lua that will catch the last 3 cases:

    debug.setmetatable(nil,{__call=function() warning("Unknown function")  end})

and it works fine when embedded a script.


I've been struggling for way too many hours now trying to convert it to C,
something like:

    // somewhere in init code

    lua_pushnil(luactx);
    lua_pushcfunction(luactx, undefined_function_handler);
    lua_setfield(luactx, -2, "__call");
    lua_setmetatable(luactx, -1);



    int undefined_function_handler(lua_State* L)
    { 
    	printf("%s is not a defined function\n", lua_getcurrentfunction(L));
    	
    	// stop execution and call error here
     
    	return 0;
    }
    
    const char* lua_getcurrentfunction(lua_State* L) {
        lua_Debug ar;
        lua_getstack(L, 1, &ar);
        lua_getinfo(L, "f", &ar);
        return ar.name;
    }

without any success.  Any ideas?

I think trapping the undeclared variables may also be useful.

Totally new to Lua I may be making this too complicated, but I think that trapping is better than crashing or providing no feedback at all.

**Comment by nm156** [comment]
I compiled openvsm-bug-11 and have started testing code again.  So far so good!

This message could be more descriptive, like "Fatal error, no pin assignments found in script".

    print_error ( this, "No device model found, it is fatal error" );

PS I dumped my playground test bed that was an MSVC 2013 C++ front end (using VSM.HPP)  and went to mingw with your entire openvsm-bug-11 code base.  

Also versioning is missing in all current branches:

    RELEASE=`../relgen.sh`

**Comment by Pugnator** [comment]
I pushed new version with much better error handling.
Undefined function check is my pain too. Will create stackoverflow question today may be.
Let's close this bug as multiple instances is already work as is, and have another issue in separate bugs. It will be easier to fix them if regress bug will appear
